addons/crm/models/crm_lead.py

```diff
@@ -1324,6 +1358,97 @@ def _pls_get_naive_bayes_probabilities(self, batch_mode=False):
 
+    # ---------------------------------
+    # PLS: Live Increment
+    # ---------------------------------
+    def _pls_reach_state(self, state):
+        """
+        When reaching won (or lost) state, by set won or set lost e.g.,
+        we need to increment the lead parameters in won (or lost) count in the frequency table,
+        to ensure the next probability computation will be based on up to date data.
+        """
+        self._pls_increment_frequencies(is_won=state == 'won')
+
+    def _pls_leave_state(self, state):
+        """
+        When leaving a won (or lost) state, when restoring a lead e.g.,
+        we need to decrement the lead parameters in won (or lost) count in the frequency table,
+        to ensure the next probability computation will be based on up to date data.
+        """
+        self._pls_increment_frequencies(is_won=state == 'won', decrement=True)
+
+    def _pls_increment_frequencies(self, is_won=True, decrement=False):
+        """
+        When losing or winning a lead, this method is called to increment each PLS parameter related to the lead
+        in won_count (if won) or in lost_count (if lost).
+
+        This method is also used when reactivating a mistakenly lost lead (using the decrement argument).
+        In this case, the lost count should be de-increment by 1 for each PLS parameter linked ot the lead.
+        """
+        pls_start_date = self._pls_get_safe_start_date()
+        if not pls_start_date:
+            return
+        # Only treat leads created after the PLS start Date
+        pls_leads = self.filtered(lambda lead: fields.Date.to_date(pls_start_date) <= fields.Date.to_date(lead.create_date))
+        if not pls_leads:
+            return
+
+        team_ids = pls_leads.mapped('team_id').ids + [0]  # If team_id is unset, consider it as team 0
+        leads_values_dict = pls_leads._pls_get_lead_pls_values()
+
+        # get current frequencies related to the target leads (currently set to won or lost)
+        # Get unique trio (field-value-team) to search in frequency table + values
+        leads_frequency_values_by_team = dict((team_id, []) for team_id in team_ids)
+        leads_values = set()
+        for lead_id, values in leads_values_dict.items():
+            team_id = values['team_id'] if values['team_id'] else 0  # If team_id is unset, consider it as team 0
+            lead_frequency_values = {'count': 1}
+            for field, value in values['values']:
+                leads_values.add((field, value, team_id))
+                if field == 'tag_id':  # handle tag_id separatelly (as in One Shot rebuild mode)
+                    leads_frequency_values_by_team[team_id].append({field: value, 'count': 1})
+                else:
+                    lead_frequency_values[field] = value
+            leads_frequency_values_by_team[team_id].append(lead_frequency_values)
+
+        leads_fields = list(set([lead_value[0] for lead_value in leads_values]))
```

Comment:
Same here, you build this weird tuple set of 3 values and only ever use the first one?

Reply:
forgot to simplify after Nieme refactoring :D